### ISSUE TITLE
Bug: [@typescript-eslint/no-confusing-void-expression] report error when the void type expression is at the left of comma operators

### ISSUE THREAD

**Issue by Leo-768** [typescript-eslint/typescript-eslint#6414]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&sourceType=module&code=GYVwdgxgLglg9mABBOcA2iAUMwAcRQBciAzlAE44DmAlIgN4CwAUIouQKZQjlKYpgS6DgDo0cKtjwEaAGkQ58UGiwC+QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90URNGgdokcGAC+IPUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
function cool (input: string) {
  return (console.log(input), input)
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-confusing-void-expression": "error"
  }
}
```

### Expected Result

No error

### Actual Result

```
Placing a void expression inside another expression is forbidden. Move it to its own statement instead. 2:11 - 2:29
```

**Comment by JoshuaKGoldberg** [comment]
...I'm confused. The rule is giving a complaint as expected. The code does indeed include a `void` expression (`console.log(input)`) inside another expression (the sequence expression). Why would this not be desirable?

**Comment by Josh-Cena** [comment]
@JoshuaKGoldberg The point of this rule is to avoid relying on the return value of a `void` expression, which could practically be anything. However, all except the last of the expressions in a sequence expression are discarded, so this is the same situation as the void expression—the entire point of the operator is to discard the operand's result.

**Comment by JoshuaKGoldberg** [comment]
That makes sense. I personally would enable a rule to ban sequence expressions, but that's out of the scope of this rule. 👍

**Comment by Josh-Cena** [comment]
[Here](https://typescript-eslint.io/play#ts=4.9.3&sourceType=module&showAST=es&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXy3TgFsRUMAKASgC54A3HLYAbgCg3EcZ4KIQMBeAF54ABhZCAPPACMEgkRCly1ADSKwJMpSpV4Abw4BfNkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90URNGgdokcGAC+IPUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) could be a more motivating example:

```ts
declare function increment(): void;

for (let i = 0; i < 10; increment(), increment()) {}
```

Even with `no-sequences`, it won't error for `for` loop afterthoughts.

**Comment by bradzacher** [comment]
Sequence expressions should never be written by a human being and are one of the worst features of the language. I will die on this hill.

The built in `no-sequences` rule is weak because it allows a lot of sequences. Restrict the syntax. Kill it all.

Legitimately i was very close to making it a syntax error in flow because people accidentally write it more than intententionally.

All that being said - this is one of the few semantically okay cases to use a void expression.

**Comment by Josh-Cena** [comment]
Well, this is probably the only non-convoluted way to provide multiple updaters in a `for` statement...

(And I can preempt Brad saying "we don't use `for` statements at Meta😉)
